packages/lwc/types.d.ts

```diff
@@ -193,17 +193,17 @@ declare module 'lwc' {
      * @param adapter the adapter used to provision data
      * @param config configuration object for the adapter
      */
-    export function wire(
-        adapter: WireAdapterConstructor | LegacyWireAdapterConstructor,
-        config?: WireConfigValue
+    export function wire<T, R>(
+        adapter: WireAdapterConstructor<T> | LegacyWireAdapterConstructor<T, R>,
+        config?: T
     ): PropertyDecorator;
 
-    type LegacyWireAdapterConstructor = (config?: any) => any;
+    type LegacyWireAdapterConstructor<T, R> = (config?: T) => R;
     type WireConfigValue = Record<string, any>;
     type ContextValue = Record<string, any>;
 
-    interface WireAdapter {
-        update(config: WireConfigValue, context?: ContextValue): void;
+    interface WireAdapter<T> {
+        update(config: T, context?: ContextValue): void;
```

Comment:
@nolanlawson, I think this one should be `config?: T extends WireConfigValue` (here on the WireAdapter)(I thought I had that already but this I missed it in my copying from a few different sources). It would only make sense to apply it on the actual `wire`'s config definition if it also applied to the `LegacyWireAdapterConstructor` as well.